### DISCUSSION TITLE
Using Directus with a NextJS front end

### DISCUSSION THREAD

**Discussion by seb3point0** [directus/directus#4981] [Archive]
Are there any good working examples of the Directus JS SDK used with a React/Next front end? Specifically, I'm looking for code examples or tutorials which use Directus Auth in order to create a user dashboard in React.

**Comment by rijkvanzanten** [comment]
Heya! The Directus SDK is the easiest way to connect to a Directus API from a JS-powered project. See https://docs.directus.io/reference/sdk/

**Reply by seb3point0** [reply]
Right, but I'm looking for existing React implementations for authenticating users.

**Reply by rijkvanzanten** [reply]
How would a React specific implementation look different from the JS SDK? 🤔

**Reply by seb3point0** [reply]
Let me rephrase that. I'm looking for code examples or tutorials explaining how to implement authentication using the SDK in React.

**Reply by fifijo** [reply]
Have a look at https://jasonwatmore.com/post/2021/08/04/next-js-11-jwt-authentication-tutorial-with-example-app, it's just an example but you will get an idea how to handle it

**Reply by joselcvarela** [reply]
Hello @seb2point0 
Here is an example from my local tests using Typescript:
```ts
import { Directus } from "@directus/sdk";
import logo from "./logo.svg";
import "./App.css";
import { useEffect, useState } from "react";

type Persons = {
  name: string;
  children: string;
}

type Schema = {
  persons: Persons;
}

const sdk = new Directus<Schema>("http://localhost:8055");

function App() {
  const [data, set_data] = useState(null);

  useEffect(() => {
    async function fetch() {
      await sdk.auth.static("s3cr3t"); // or you can use await directus.auth.login({ email: 'admin@example.com', password: 'd1r3ctu5' });
      

      const result = await sdk.items("persons").readMany();
      
      set_data(result);
    }

    fetch();
  }, []);

  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
      {do_something_with_data(data)}
    </div>
  );
}

export default App;
```

**Reply by seb3point0** [reply]
> Have a look at https://jasonwatmore.com/post/2021/08/04/next-js-11-jwt-authentication-tutorial-with-example-app, it's just an example but you will get an idea how to handle it

Thanks.  What I can't seem to find is a tutorial of this quality explaining how to use Directus' Auth API. Perhaps it doesn't exist yet.

**Reply by alsherif-khalaf** [reply]
> > Have a look at https://jasonwatmore.com/post/2021/08/04/next-js-11-jwt-authentication-tutorial-with-example-app, it's just an example but you will get an idea how to handle it
> 
> Thanks. What I can't seem to find is a tutorial of this quality explaining how to use Directus' Auth API. Perhaps it doesn't exist yet.

Hello
Did you find any tutorial for the same purpose ?

**Comment by raphourbe** [comment]
Hi, I have also been struggling with this and I might I've found a way.

I actually had one main problem at the beginning: having several useSWR functions calling the directus API at the same time generating several refresh token calls. It ended up in making my whole app bug. So I figured a way to leverage the directus sdk to manage the refresh token part + the authentication. 

It is highly inspired by this example that I have adapted (https://www.mikealche.com/software-development/how-to-implement-authentication-in-next-js-without-third-party-libraries)

I have a directus.js file that is quite simple:
```javascript

// directus.js
import { Directus } from "@directus/sdk";

const directus = new Directus(process.env.NEXT_PUBLIC_API);

export default directus;

```

Then I have modified an AuthProvider component like this:

```javascript

// file: auth.js
import React, { createContext, useState, useContext, useEffect } from "react";

// directus api
import directus from "../api/directus";

const AuthContext = createContext({});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [loading, setLoading] = useState(true);
  const isAuthenticated = !!user;

  useEffect(() => {
    async function loadUserFromDirectus() {
      const token = await directus.auth.token;
      if (token) {
        const me = await directus.users.me.read();
        setUser(me);
      }
      setLoading(false);
    }
    setLoading(true);
    loadUserFromDirectus();
  }, []);

  const login = async ({ email, password }) => {
    setLoading(true);
    let authenticated;
    const res = await directus.auth
      .login({
        email: email,
        password: password,
      })
      .then(() => {
        authenticated = true;
      })
      .catch((error) => {
        console.log(error);
        return error;
      });
    if (res?.errors?.length > 0) return res;

    const me = await directus.users.me.read();
    setUser(me);
    setLoading(false);
  };

  const logout = async () => {
    await directus.auth.logout();
    setUser(null);
  };

  const getToken = async () => {
    await directus.auth.refreshIfExpired().catch((error) => {
      setUser(null);
    });
    const token = await directus.auth.token;
    if (token) {
      return token;
    }
    setUser(null);
    return null;
  };

  return (
    <AuthContext.Provider
      value={{ isAuthenticated, user, login, loading, logout, getToken }}
    >
      {children}
    </AuthContext.Provider>
  );
};
export const useAuth = () => useContext(AuthContext);
```

I have added a getToken() function that can use the refreshIfExpired() function of directus. So when I use multiple useSWR() to retrieve data from directus (I am not always using the sdk to retrieve data) I don't have trouble with refreshing token. If no token is retrieved, then user becomes null, so the user is kicked out of the app.

My _app.js file is like:

```javascript
import React from "react";
import { SWRConfig } from "swr";
import "../styles/globals.css";
import { AuthProvider } from "../lib/auth";

function MyApp({ Component, pageProps: { ...pageProps } }) {
  return (
    <AuthProvider>
      <SWRConfig
        value={{
          revalidateOnMount: true,
        }}
      >
        <Component {...pageProps} />
      </SWRConfig>
    </AuthProvider>
    // </SessionProvider>
  );
}

export default MyApp;
```

In a component using a useSWR function it goes like this: 
```javascript

import { useAuth } from "../../../lib/auth";
import useSWR from "swr";
import fetchData from "./fetcherHelper";

export default function Publication({ profile }) {
  const { getToken } = useAuth();
  const { actualCompo, error, isLoading } = usePublication (
    getToken,
    profile
  );
  if (error) {
    if (error.status === 401)
      return "Loading";
    return "Error";
  }
  else if (isLoading) return "Loading.";
  return <div>Something</div>;
}

const usePortfolioActualCompo = (getToken, profile) => {
  const { data, error } = useSWR(
    "publication/" + profile,
    (query) => {
      return fetchData(query, getToken,"publication/" + profile);
    },
    {
      onErrorRetry: (error, key, config, revalidate, { retryCount }) => {
        // Never retry on 404.
        if (error.status === 404) return;

        // Only retry up to 5 times.
        if (retryCount >= 5) return;

        // Retry after 1 seconds.
        setTimeout(() => revalidate({ retryCount }), 1000);
      },
      refreshInterval: 15 * 60 * 1000,
      revalidateOnFocus: true,
    }
  );

  return {
    actualCompo: data,
    error: error,
    isLoading: !data && !error,
  };
};
```

My fetchData function looks like this. I use the getToken() function in it. So everytime I use my fetcher, it checks if it needs to refresh the token.

```javascript

// fetchHelper.js
const pubAPI = process.env.NEXT_PUBLIC_API;

const fetchData = async (query, getToken, urlAdd, variables) => {
  const token = await getToken();
  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${token}`,
  };

  var urlAPI = pubAPI;
  if (urlAdd !== undefined) urlAPI = urlAPI + urlAdd;

  const res = await fetch(urlAPI, {
    method: "POST",
    headers,
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  if (!res.ok) {
    const error = new Error("An error occurred while fetching the data.");
    // Attach extra info to the error object.
    error.info = await res.json();
    error.status = res.status;
    throw error;
  }

  const json = await res.json();

  return json;
};

export default fetchData;
```

I have set in the directus .env file the value: ACCESS_TOKEN_TTL="1m" to test this solution. Since several useSWR functions fire all at the same time, some of them get an error 401 while the first one refresh the token. Using the revalidate function in useSWR on error allows you to retry quickly enough and refetch with a correct token. I added a "loading" indicator when I get a 401 response from the server while refreshing token. If the token has totally expired, the catch after refrechIfExpired() sets the user value in the AuthProvider becomes null. I have a verification that user is always connected, so it will kick the user out of the app if this case happens.

```javascript

import React, { Children, cloneElement, useEffect } from "react";
import { useAuth } from "../../lib/auth";
import { useRouter } from "next/router.js";

export default function SessionLayout({ children }) {
  const router = useRouter();
  const { user } = useAuth();
  useEffect(() => {
    if (!user) router.push("/login");
  }, [user, router]);

  if (user) {
    return (
      <Layout>
        {Children.map(children, (child) => {
          return cloneElement(child, {
            ...child.props,
            user: user,
          });
        })}
      </Layout>
    );
  } else {
    return "Nothing";
  }
}
```

I hope this solution helps people. I am not a professional dev so if there is some problems here, I would be very happy to have information regarding this topic 👍

**Reply by BDive** [reply]
Thank you for the detailed example, @raphourbe . Using the above example, I noticed that the consumer (child components) only receive an updated context value on every second re-render. 

As an example, If I return setState function from AuthProvider and call that from any child component to update state, the component re-renders but it returns previous value until re-rendered second time. I attempted to examine the code but was unsuccessful (I am currently learning react). Any ideas?  

Sample code of a child component

```javascript

import React from 'react';
import { useAuth } from './auth';


export default function TestPage() {

    const {loading, setLoading} = useAuth(); //setLoading returned from AuthProvider in auth.js, loading is set to string for testing


    const clickHandler = () => {
        console.log(`Before :${loading}`);
        setLoading(document.getElementById('test').value);
        console.log(`After :${loading}`);
    }

    return(
    <div>
         <input type='text' id="test" />
        <br />
        <button onClick={clickHandler}>Click</button>
    </div>
    );
}

```

**Reply by BDive** [reply]
After a good night's sleep, I was able to pinpoint the source of the problem :) I guess it is a problem for many react beginners. The problem is that useState executes asynchronously, so we don't get instant results. I'm leaving comment on here just in case it might help someone with a similar problem.

**Comment by jomuji** [comment]
Anyone have tried this with Google provider?
